Add profiling support to the emulator.

jdeeny opened the pull request:
Each time an instruction is decoded, a count of ticks spent at the current PC is incremented. The user can press 'P' to show profiling information. The per-PC data is reduced into blocks of code
based on shared labels, so the granularity can be increased by adding additional labels to the code.

Please let me know if you would prefer any part of this done differently.

jdeeny committed:
Add profiling support to the emulator.

Each time an instruction is decoded, a count of tick spent
at the current PC is incremented. The user can press 'P' to
show profiling information.

jdeeny commented:
Example: 
![alt text](http://i.imgur.com/Z9caNDl.png)

JohnEarnest commented:
This definitely seems like a handy feature! I think the way you're presenting the information is a bit overwhelming. How about something like the following:
- Limit displayed results to the highest-weighted 20 entries or so. If somebody really wants all the gory detail, they can refer to the textbox where you dump all the information you gather as JSON for external analysis. The way things are laid out presently, a profile of a large program will fill most screens and render that textarea inaccessible. I think most of the time users will just want a quick summary.
- Cut down what you display to just the essentials: ticks spent in a region, percentage of execution time and the symbolic name and extent of the region. Summaries could look something like this:

```
ticks   time    source
408     22.37%  (random-vector) +20
148     8.11%   (fill-boards) +10
80      4.39%   (restock-single + 26) +6
316     17.32%  (map-tile) +34
204     11.18%  (restock-single) +22
228     12.50%  (place-owned-single) +26
139     7.62%   (draw-map) +28
76      4.17%   (restock) +24
76      4.17%   (place-owned) +24
```

Better use of screen real-estate, but all the most important details preserved.

What do you think?

jdeeny committed:
Cleanup output

jdeeny committed:
Improve JSON output

jdeeny commented:
I like your output format, much nicer. I changed the output to match your example. I also added the calls column, showing how many times execution has passed through this block (or at least the first instruction).

![example](http://i.imgur.com/kdO4Tzw.png)

JohnEarnest commented:
Looks great! Thanks for taking the time to work this up.
